File: android/app/src/main/java/com/vpn/android/vpn/xray/VlessUri.java
package com.vpn.android.vpn.xray;

import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Parses the {@code vless://uuid@host:port?query#remark} links produced by
 * SubscriptionExportService.buildVlessUrl on the server. Query parameter
 * names match that method's output: encryption, security, type, path, sni,
 * pbk, sid.
 */
public final class VlessUri {

    private final String uuid;
    private final String host;
    private final int port;
    private final Map<String, String> params;
    private final String remark;

    private VlessUri(String uuid, String host, int port, Map<String, String> params, String remark) {
        this.uuid = uuid;
        this.host = host;
        this.port = port;
        this.params = params;
        this.remark = remark;
    }

    public static VlessUri parse(String link) {
        if (link == null || !link.startsWith("vless://")) {
            throw new IllegalArgumentException("Not a vless:// link: " + link);
        }
        URI uri;
        try {
            uri = new URI(link);
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("Malformed vless link: " + link, e);
        }

        String uuid = uri.getUserInfo();
        String host = uri.getHost();
        int port = uri.getPort();
        if (uuid == null || uuid.isEmpty() || host == null || port <= 0) {
            throw new IllegalArgumentException("Incomplete vless link: " + link);
        }

        Map<String, String> params = parseQuery(uri.getRawQuery());
        String remark = decode(uri.getRawFragment());

        return new VlessUri(uuid, host, port, params, remark != null ? remark : "");
    }

    private static Map<String, String> parseQuery(String rawQuery) {
        Map<String, String> result = new LinkedHashMap<>();
        if (rawQuery == null || rawQuery.isEmpty()) {
            return result;
        }
        for (String pair : rawQuery.split("&")) {
            if (pair.isEmpty()) continue;
            int eq = pair.indexOf('=');
            String key = eq >= 0 ? pair.substring(0, eq) : pair;
            String value = eq >= 0 ? pair.substring(eq + 1) : "";
            result.put(decode(key), decode(value));
        }
        return result;
    }

    private static String decode(String value) {
        if (value == null) return null;
        try {
            return URLDecoder.decode(value, StandardCharsets.UTF_8.name());
        } catch (UnsupportedEncodingException e) {
            return value;
        }
    }

    public String getUuid() {
        return uuid;
    }

    public String getHost() {
        return host;
    }

    public int getPort() {
        return port;
    }

    public String getParam(String key, String defaultValue) {
        String v = params.get(key);
        return v != null ? v : defaultValue;
    }

    public Map<String, String> getParams() {
        return Collections.unmodifiableMap(params);
    }

    public String getRemark() {
        return remark;
    }
}

File: desktop/src/shared/vlessUri.ts
/**
 * Parses the vless://uuid@host:port?query#remark links produced by
 * server SubscriptionExportService#buildVlessUrl. Query parameter names
 * match that method's output: encryption, security, type, path, sni, pbk, sid.
 */
export interface ParsedVlessUri {
  uuid: string;
  host: string;
  port: number;
  params: Record<string, string>;
  remark: string;
}

export function parseVlessUri(link: string): ParsedVlessUri {
  if (!link || !link.startsWith('vless://')) {
    throw new Error(`Not a vless:// link: ${link}`);
  }

  let url: URL;
  try {
    url = new URL(link);
  } catch (e) {
    throw new Error(`Malformed vless link: ${link}`);
  }

  const uuid = decodeURIComponent(url.username);
  const host = url.hostname;
  const port = url.port ? Number(url.port) : 0;
  if (!uuid || !host || !port) {
    throw new Error(`Incomplete vless link: ${link}`);
  }

  const params: Record<string, string> = {};
  url.searchParams.forEach((value, key) => {
    params[key] = value;
  });

  const remark = url.hash ? decodeURIComponent(url.hash.slice(1)) : '';

  return { uuid, host, port, params, remark };
}

export function vlessParam(uri: ParsedVlessUri, key: string, fallback: string): string {
  return uri.params[key] ?? fallback;
}

/**
 * The user-facing half of a link's remark: the server builds it as
 * "<region> · <node hostname>" (SubscriptionExportService#REMARK_SEPARATOR).
 * Only the region belongs on screen — the hostname is operator detail, and
 * with the old "-" separator it leaked into the UI as part of the region name.
 */
export function regionLabel(remark: string): string {
  const [region] = remark.split(' \u00b7 ');
  return (region || remark).trim();
}

/**
 * The first link that actually belongs to `region`, or null if none does.
 *
 * Used to ping the selected region: the server falls back to any online node
 * when the asked-for region has none (`requestedRegionAvailable: false`), so
 * taking whatever link came back would report a node in a different country as
 * this region's latency. Malformed links are skipped rather than throwing —
 * one bad entry should not cost the measurement.
 */
export function firstLinkForRegion(links: string[], region: string): ParsedVlessUri | null {
  for (const link of links) {
    try {
      const parsed = parseVlessUri(link);
      if (parsed.remark && regionLabel(parsed.remark) === region) {
        return parsed;
      }
    } catch {
      // skip a malformed link
    }
  }
  return null;
}
